crates/fs/private/node.rs

```diff
@@ -0,0 +1,280 @@
+
+use super::{
+    hamt::Hasher, namefilter::Namefilter, Key, PrivateDirectory, PrivateDirectoryContent,
+    PrivateFile, PrivateFileContent,
+};
+
+//--------------------------------------------------------------------------------------------------
+// Type Definitions
+//--------------------------------------------------------------------------------------------------
+
+pub type INumber = HashOutput;
+
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct ContentKey(pub Key);
+
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct RatchetKey(pub Key);
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct PrivateNodeHeader {
+    pub(crate) bare_name: Namefilter,
+    pub(crate) ratchet: Ratchet,
+    pub(crate) inumber: INumber,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum PrivateNode {
+    File(Rc<PrivateFile>),
+    Dir(Rc<PrivateDirectory>),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct PrivateRef {
+    pub(crate) saturated_name_hash: HashOutput, // Sha3-256 hash of saturated namefilter
+    pub(crate) content_key: ContentKey,         // A hash of ratchet key.
+    pub(crate) ratchet_key: RatchetKey,         // Encrypted ratchet key.
+}
+
+//--------------------------------------------------------------------------------------------------
+// Implementations
+//--------------------------------------------------------------------------------------------------
+
+impl PrivateNodeHeader {
+    /// Creates a new PrivateNodeHeader.
+    pub fn new(parent_bare_name: Namefilter, inumber: INumber, ratchet_seed: HashOutput) -> Self {
+        Self {
+            bare_name: {
+                let mut namefilter = parent_bare_name;
+                namefilter.add(&inumber);
+                namefilter
+            },
+            ratchet: Ratchet::zero(ratchet_seed),
+            inumber,
+        }
+    }
+
+    /// Advances the ratchet.
+    pub fn advance_ratchet(&mut self) {
+        self.ratchet.inc();
+    }
+
+    /// Gets the private ref of the current header.
+    pub fn get_private_ref(&self) -> Result<PrivateRef> {
+        let ratchet_key = Key::new(self.ratchet.derive_key());
+        let saturated_name_hash = {
+            let mut name = self.bare_name.clone();
+            name.add(&ratchet_key.as_bytes());
+            name.saturate();
+            Sha3_256::hash(&name.as_bytes())
+        };
+
+        Ok(PrivateRef {
+            saturated_name_hash,
+            content_key: ContentKey(Key::new(Sha3_256::hash(&ratchet_key.as_bytes()))),
+            ratchet_key: RatchetKey(ratchet_key),
+        })
+    }
+
+    /// Gets the saturated namefilter for this node.
+    pub fn get_saturated_name(&self) -> Namefilter {
+        let ratchet_key = Key::new(self.ratchet.derive_key());
+        let mut name = self.bare_name.clone();
+        name.add(&ratchet_key.as_bytes());
+        name.saturate();
+        name
+    }
```

Comment:
Normally I wouldn't care about repeating some code *once*, but in this case that's a pretty crucial piece of code.
Let's extract the common part of these two functions into a third, private one.

(Also, that was me who duplicated that code when we were pairing :P Sorry!)